File: dlitscustom/override/sales_invoice_commission.js
// Copyright (c) 2024, DLITS and contributors
// For license information, please see license.txt

// Sales Invoice Commission JavaScript

frappe.ui.form.on('Sales Invoice', {
    refresh: function(frm) {
        // Add commission-related buttons
        add_commission_buttons(frm);
        
        // Set field properties
        set_commission_field_properties(frm);
        
        // Update commission display
        update_commission_display(frm);
    },
    
    dlits_sales_partner: function(frm) {
        // When sales partner is selected, get commission details and calculate
        if (frm.doc.dlits_sales_partner) {
            get_sales_partner_commission_details(frm);
        } else {
            clear_commission_fields(frm);
        }
    },
    
    grand_total: function(frm) {
        // Recalculate commission when grand total changes
        if (frm.doc.dlits_sales_partner && frm.doc.grand_total) {
            calculate_commission_amount(frm);
        }
    },
    
    dlits_commission_paid_amount: function(frm) {
        // Update outstanding commission when paid amount changes
        update_outstanding_commission(frm);
    }
});

function add_commission_buttons(frm) {
    if (!frm.is_new() && frm.doc.docstatus === 1) {
        
        // Update Commission Paid button
        if (frm.doc.dlits_commission_outstanding > 0) {
            frm.add_custom_button(__('Update Commission Paid'), function() {
                update_commission_paid_dialog(frm);
            }, __('Commission'));
        }
        
        // Create Commission Payment button
        if (frm.doc.dlits_commission_outstanding > 0 && frm.doc.dlits_sales_partner) {
            frm.add_custom_button(__('Create Commission Payment'), function() {
                create_commission_payment(frm);
            }, __('Commission'));
        }
        
        // View Commission Report button
        if (frm.doc.dlits_sales_partner) {
            frm.add_custom_button(__('View Commission Report'), function() {
                view_commission_report(frm);
            }, __('Commission'));
        }
    }
}

function set_commission_field_properties(frm) {
    // Set commission fields as read-only based on conditions
    frm.set_df_property('dlits_commission_type', 'read_only', 1);
    frm.set_df_property('dlits_commission_rate', 'read_only', 1);
    frm.set_df_property('dlits_commission_amount', 'read_only', 1);
    frm.set_df_property('dlits_commission_outstanding', 'read_only', 1);
    
    // Commission paid amount can be edited after submit
    if (frm.doc.docstatus === 1) {
        frm.set_df_property('dlits_commission_paid_amount', 'read_only', 0);
    }
}

function get_sales_partner_commission_details(frm) {
    if (!frm.doc.dlits_sales_partner) return;
    
    frappe.call({
        method: 'dlitscustom.override.sales_invoice_commission.get_sales_partner_details',
        args: {
            sales_partner: frm.doc.dlits_sales_partner
        },
        callback: function(r) {
            if (r.message) {
                let details = r.message;
                
                // Check if partner is active
                if (details.status !== 'Active') {
                    frappe.msgprint(__('Selected sales partner is not active'));
                    frm.set_value('dlits_sales_partner', '');
                    return;
                }
                
                // Set commission details
                frm.set_value('dlits_commission_type', details.commission_type);
                frm.set_value('dlits_commission_rate', details.commission_rate);
                
                // Calculate commission amount
                calculate_commission_amount(frm);
            }
        }
    });
}

function calculate_commission_amount(frm) {
    if (!frm.doc.dlits_sales_partner || !frm.doc.grand_total) return;
    
    frappe.call({
        method: 'dlitscustom.override.sales_invoice_commission.calculate_commission_amount',
        args: {
            sales_partner: frm.doc.dlits_sales_partner,
            grand_total: frm.doc.grand_total
        },
        callback: function(r) {
            if (r.message) {
                frm.set_value('dlits_commission_amount', r.message);
                update_outstanding_commission(frm);
            }
        }
    });
}

function update_outstanding_commission(frm) {
    let outstanding = flt(frm.doc.dlits_commission_amount) - flt(frm.doc.dlits_commission_paid_amount);
    frm.set_value('dlits_commission_outstanding', outstanding);
}

function clear_commission_fields(frm) {
    frm.set_value('dlits_commission_type', '');
    frm.set_value('dlits_commission_rate', 0);
    frm.set_value('dlits_commission_amount', 0);
    frm.set_value('dlits_commission_paid_amount', 0);
    frm.set_value('dlits_commission_outstanding', 0);
}

function update_commission_paid_dialog(frm) {
    let dialog = new frappe.ui.Dialog({
        title: __('Update Commission Paid Amount'),
        fields: [
            {
                label: 'Commission Amount',
                fieldname: 'commission_amount',
                fieldtype: 'Currency',
                default: frm.doc.dlits_commission_amount,
                read_only: 1
            },
            {
                label: 'Currently Paid',
                fieldname: 'current_paid',
                fieldtype: 'Currency',
                default: frm.doc.dlits_commission_paid_amount,
                read_only: 1
            },
            {
                label: 'New Paid Amount',
                fieldname: 'new_paid_amount',
                fieldtype: 'Currency',
                default: frm.doc.dlits_commission_paid_amount,
                reqd: 1
            },
            {
                label: 'Notes',
                fieldname: 'notes',
                fieldtype: 'Small Text',
                description: 'Optional notes for this payment update'
            }
        ],
        primary_action_label: __('Update'),
        primary_action: function(values) {
            if (flt(values.new_paid_amount) > flt(frm.doc.dlits_commission_amount)) {
                frappe.msgprint(__('Paid amount cannot be greater than commission amount'));
                return;
            }
            
            if (flt(values.new_paid_amount) < 0) {
                frappe.msgprint(__('Paid amount cannot be negative'));
                return;
            }
            
            frappe.call({
                method: 'dlitscustom.override.sales_invoice_commission.update_commission_paid_amount',
                args: {
                    invoice_name: frm.doc.name,
                    paid_amount: values.new_paid_amount
                },
                callback: function(r) {
                    if (r.message && r.message.success) {
                        frappe.msgprint(r.message.message);
                        frm.reload_doc();
                    } else if (r.message) {
                        frappe.msgprint(r.message.message);
                    }
                }
            });
            
            dialog.hide();
        }
    });
    
    dialog.show();
}

function create_commission_payment(frm) {
    let outstanding = frm.doc.dlits_commission_outstanding;
    
    frappe.prompt([
        {
            label: 'Payment Amount',
            fieldname: 'amount',
            fieldtype: 'Currency',
            default: outstanding,
            reqd: 1
        },
        {
            label: 'Reference',
            fieldname: 'reference',
            fieldtype: 'Data',
            description: 'Optional reference for this payment'
        }
    ], function(values) {
        frappe.call({
            method: 'dlitscustom.dlitscustom.doctype.dlits_sales_partner.dlits_sales_partner.create_commission_payment',
            args: {
                sales_partner: frm.doc.dlits_sales_partner,
                amount: values.amount,
                reference_doc: frm.doc.name
            },
            callback: function(r) {
                if (r.message && r.message.payment_entry) {
                    frappe.msgprint(__('Payment entry created successfully'));
                    
                    // Ask if user wants to open the payment entry
                    frappe.confirm(__('Do you want to open the payment entry?'), function() {
                        frappe.set_route('Form', 'Payment Entry', r.message.payment_entry.name);
                    });
                    
                    // Update the paid amount in this invoice
                    frappe.call({
                        method: 'dlitscustom.override.sales_invoice_commission.update_commission_paid_amount',
                        args: {
                            invoice_name: frm.doc.name,
                            paid_amount: flt(frm.doc.dlits_commission_paid_amount) + flt(values.amount)
                        },
                        callback: function(r) {
                            if (r.message && r.message.success) {
                                frm.reload_doc();
                            }
                        }
                    });
                }
            }
        });
    }, __('Create Commission Payment'), __('Create'));
}

function view_commission_report(frm) {
    // Use direct URL to avoid getdoctype issues
    let url = `/app/query-report/DLITS%20Sales%20Partner%20Commission%20Report?sales_partner=${encodeURIComponent(frm.doc.dlits_sales_partner)}&from_date=${frm.doc.posting_date}&to_date=${frm.doc.posting_date}`;
    window.open(url, '_blank');
}

function update_commission_display(frm) {
    // Add visual indicators for commission status
    if (frm.doc.dlits_commission_outstanding > 0) {
        frm.dashboard.add_indicator(__('Outstanding Commission: {0}', [format_currency(frm.doc.dlits_commission_outstanding)]), 'orange');
    } else if (frm.doc.dlits_commission_amount > 0 && frm.doc.dlits_commission_outstanding === 0) {
        frm.dashboard.add_indicator(__('Commission Fully Paid'), 'green');
    }
    
    if (frm.doc.dlits_sales_partner) {
        frm.dashboard.add_indicator(__('Sales Partner: {0}', [frm.doc.dlits_sales_partner]), 'blue');
    }
}

// Utility function to format currency
function format_currency(amount) {
    return frappe.format(amount, {fieldtype: 'Currency'});
}

// Auto-calculate commission on form load if sales partner is set
frappe.ui.form.on('Sales Invoice', {
    onload: function(frm) {
        if (frm.doc.dlits_sales_partner && !frm.doc.dlits_commission_amount) {
            get_sales_partner_commission_details(frm);
        }
    }
});

File: dlitscustom/override/sales_invoice_commission.py
# Copyright (c) 2024, DLITS and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import flt

def validate_sales_invoice_commission(doc, method):
    """Validate and calculate commission for Sales Invoice"""
    if doc.dlits_sales_partner:
        calculate_commission(doc)
        validate_commission_fields(doc)

def on_submit_sales_invoice_commission(doc, method):
    """Update sales partner totals when invoice is submitted"""
    if doc.dlits_sales_partner and doc.dlits_commission_amount:
        update_sales_partner_totals(doc.dlits_sales_partner)

def on_cancel_sales_invoice_commission(doc, method):
    """Update sales partner totals when invoice is cancelled"""
    if doc.dlits_sales_partner and doc.dlits_commission_amount:
        update_sales_partner_totals(doc.dlits_sales_partner)

def calculate_commission(doc):
    """Calculate commission amount based on sales partner configuration or manual input"""
    if not doc.dlits_sales_partner:
        return
    
    try:
        # Get sales partner details
        sales_partner = frappe.get_doc("DLITS Sales Partner", doc.dlits_sales_partner)
        
        # Check if partner is active
        if sales_partner.status != "Active":
            frappe.throw(f"Sales Partner {sales_partner.partner_name} is not active")
        
        # Set commission type from sales partner (default)
        if not doc.dlits_commission_type:
            doc.dlits_commission_type = sales_partner.commission_type
        
        # Calculate commission based on manual input or default rate
        base_amount = flt(doc.grand_total)
        commission_amount = 0
        
        # If commission amount is manually entered, calculate rate
        if doc.dlits_commission_amount and not doc.dlits_commission_rate:
            commission_amount = flt(doc.dlits_commission_amount)
            if base_amount > 0:
                doc.dlits_commission_rate = (commission_amount / base_amount) * 100
        
        # If commission rate is manually entered, calculate amount
        elif doc.dlits_commission_rate:
            if doc.dlits_commission_type == "Percentage":
                commission_amount = base_amount * flt(doc.dlits_commission_rate) / 100
            else:  # Fixed Amount
                commission_amount = flt(doc.dlits_commission_rate)
            doc.dlits_commission_amount = commission_amount
        
        # If neither is entered, use sales partner defaults
        else:
            doc.dlits_commission_rate = sales_partner.commission_rate
            if sales_partner.commission_type == "Percentage":
                commission_amount = base_amount * flt(sales_partner.commission_rate) / 100
            else:  # Fixed Amount
                commission_amount = flt(sales_partner.commission_rate)
            doc.dlits_commission_amount = commission_amount
        
        # Apply minimum and maximum limits (only if using default calculation)
        if not (doc.dlits_commission_amount and doc.dlits_commission_rate != sales_partner.commission_rate):
            if sales_partner.minimum_commission and commission_amount < flt(sales_partner.minimum_commission):
                commission_amount = flt(sales_partner.minimum_commission)
                doc.dlits_commission_amount = commission_amount
            
            if sales_partner.maximum_commission and commission_amount > flt(sales_partner.maximum_commission):
                commission_amount = flt(sales_partner.maximum_commission)
                doc.dlits_commission_amount = commission_amount
        
        # Calculate outstanding commission
        doc.dlits_commission_outstanding = flt(doc.dlits_commission_amount) - flt(doc.dlits_commission_paid_amount or 0)
        
        # Log commission calculation
        frappe.logger().info(f"Commission calculated for {doc.name}: {commission_amount} for partner {sales_partner.partner_name}")
        
    except Exception as e:
        frappe.log_error(f"Error calculating commission for {doc.name}: {str(e)}")
        frappe.throw(f"Error calculating commission: {str(e)}")

def validate_commission_fields(doc):
    """Validate commission-related fields"""
    if doc.dlits_commission_paid_amount and flt(doc.dlits_commission_paid_amount) > flt(doc.dlits_commission_amount):
        frappe.throw("Commission paid amount cannot be greater than commission amount")
    
    if doc.dlits_commission_paid_amount and flt(doc.dlits_commission_paid_amount) < 0:
        frappe.throw("Commission paid amount cannot be negative")

def update_sales_partner_totals(sales_partner_name):
    """Update sales partner commission totals"""
    try:
        sales_partner = frappe.get_doc("DLITS Sales Partner", sales_partner_name)
        sales_partner.update_commission_totals()
        frappe.logger().info(f"Updated commission totals for {sales_partner_name}")
    except Exception as e:
        frappe.log_error(f"Error updating sales partner totals for {sales_partner_name}: {str(e)}")

@frappe.whitelist()
def get_sales_partner_details(sales_partner):
    """Get sales partner details for commission calculation"""
    if not sales_partner:
        return {}
    
    try:
        partner = frappe.get_doc("DLITS Sales Partner", sales_partner)
        return {
            "commission_type": partner.commission_type,
            "commission_rate": partner.commission_rate,
            "minimum_commission": partner.minimum_commission,
            "maximum_commission": partner.maximum_commission,
            "status": partner.status
        }
    except Exception as e:
        frappe.log_error(f"Error getting sales partner details: {str(e)}")
        return {}

@frappe.whitelist()
def calculate_commission_amount(sales_partner, grand_total):
    """Calculate commission amount for given parameters"""
    if not sales_partner or not grand_total:
        return 0
    
    try:
        partner = frappe.get_doc("DLITS Sales Partner", sales_partner)
        base_amount = flt(grand_total)
        
        if partner.commission_type == "Percentage":
            commission = base_amount * flt(partner.commission_rate) / 100
        else:  # Fixed Amount
            commission = flt(partner.commission_rate)
        
        # Apply limits
        if partner.minimum_commission and commission < flt(partner.minimum_commission):
            commission = flt(partner.minimum_commission)
        
        if partner.maximum_commission and commission > flt(partner.maximum_commission):
            commission = flt(partner.maximum_commission)
        
        return commission
        
    except Exception as e:
        frappe.log_error(f"Error calculating commission amount: {str(e)}")
        return 0

@frappe.whitelist()
def update_commission_paid_amount(invoice_name, paid_amount):
    """Update commission paid amount for an invoice"""
    try:
        doc = frappe.get_doc("Sales Invoice", invoice_name)
        
        if flt(paid_amount) > flt(doc.dlits_commission_amount):
            frappe.throw("Paid amount cannot be greater than commission amount")
        
        if flt(paid_amount) < 0:
            frappe.throw("Paid amount cannot be negative")
        
        doc.db_set("dlits_commission_paid_amount", flt(paid_amount))
        doc.db_set("dlits_commission_outstanding", flt(doc.dlits_commission_amount) - flt(paid_amount))
        
        # Update sales partner totals
        if doc.dlits_sales_partner:
            update_sales_partner_totals(doc.dlits_sales_partner)
        
        return {
            "success": True,
            "message": f"Commission paid amount updated to {paid_amount}",
            "outstanding": flt(doc.dlits_commission_amount) - flt(paid_amount)
        }
        
    except Exception as e:
        frappe.log_error(f"Error updating commission paid amount: {str(e)}")
        return {
            "success": False,
            "message": str(e)
        }

@frappe.whitelist()
def get_unpaid_commission_invoices(sales_partner=None, from_date=None, to_date=None):
    """Get invoices with unpaid commission"""
    conditions = ["docstatus = 1", "dlits_commission_outstanding > 0"]
    values = []
    
    if sales_partner:
        conditions.append("dlits_sales_partner = %s")
        values.append(sales_partner)
    
    if from_date:
        conditions.append("posting_date >= %s")
        values.append(from_date)
    
    if to_date:
        conditions.append("posting_date <= %s")
        values.append(to_date)
    
    where_clause = " AND ".join(conditions)
    
    invoices = frappe.db.sql(f"""
        SELECT 
            name,
            posting_date,
            customer,
            dlits_sales_partner,
            grand_total,
            dlits_commission_amount,
            dlits_commission_paid_amount,
            dlits_commission_outstanding
        FROM `tabSales Invoice`
        WHERE {where_clause}
        ORDER BY posting_date DESC
    """, values, as_dict=True)
    
    return invoices

@frappe.whitelist()
def create_bulk_commission_payment(invoices, sales_partner):
    """Create bulk commission payment for multiple invoices"""
    try:
        if not invoices or not sales_partner:
            frappe.throw("Invoices and sales partner are required")
        
        invoice_list = frappe.parse_json(invoices) if isinstance(invoices, str) else invoices
        
        total_amount = 0
        invoice_names = []
        
        for invoice_name in invoice_list:
            invoice = frappe.get_doc("Sales Invoice", invoice_name)
            if invoice.dlits_sales_partner == sales_partner and invoice.dlits_commission_outstanding > 0:
                total_amount += flt(invoice.dlits_commission_outstanding)
                invoice_names.append(invoice_name)
        
        if total_amount == 0:
            frappe.throw("No outstanding commission found for selected invoices")
        
        # Create payment entry
        partner = frappe.get_doc("DLITS Sales Partner", sales_partner)
        payment_entry = partner.create_commission_payment_entry(
            total_amount, 
            f"Bulk payment for invoices: {', '.join(invoice_names)}"
        )
        
        return {
            "success": True,
            "payment_entry": payment_entry.name,
            "amount": total_amount,
            "invoices": invoice_names
        }
        
    except Exception as e:
        frappe.log_error(f"Error creating bulk commission payment: {str(e)}")
        return {
            "success": False,
            "message": str(e)
        }

@frappe.whitelist()
def adjust_commission_after_submit(invoice_name, commission_amount=None, commission_rate=None):
    """Allow manual adjustment of commission after invoice submission"""
    try:
        doc = frappe.get_doc("Sales Invoice", invoice_name)
        
        if doc.docstatus != 1:
            frappe.throw("Invoice must be submitted to adjust commission")
        
        if not doc.dlits_sales_partner:
            frappe.throw("No DLITS Sales Partner assigned to this invoice")
        
        # Get sales partner details
        sales_partner = frappe.get_doc("DLITS Sales Partner", doc.dlits_sales_partner)
        base_amount = flt(doc.grand_total)
        
        # Calculate new commission based on input
        if commission_amount:
            new_commission_amount = flt(commission_amount)
            if base_amount > 0:
                new_commission_rate = (new_commission_amount / base_amount) * 100
            else:
                new_commission_rate = 0
        elif commission_rate:
            new_commission_rate = flt(commission_rate)
            if doc.dlits_commission_type == "Percentage":
                new_commission_amount = base_amount * new_commission_rate / 100
            else:
                new_commission_amount = new_commission_rate
        else:
            frappe.throw("Either commission amount or commission rate must be provided")
        
        # Validate the new values
        if new_commission_amount < 0:
            frappe.throw("Commission amount cannot be negative")
        
        if new_commission_rate < 0:
            frappe.throw("Commission rate cannot be negative")
        
        # Update the invoice
        old_commission = flt(doc.dlits_commission_amount)
        doc.db_set("dlits_commission_amount", new_commission_amount)
        doc.db_set("dlits_commission_rate", new_commission_rate)
        
        # Recalculate outstanding commission
        outstanding = new_commission_amount - flt(doc.dlits_commission_paid_amount or 0)
        doc.db_set("dlits_commission_outstanding", outstanding)
        
        # Update sales partner totals
        update_sales_partner_totals(doc.dlits_sales_partner)
        
        # Log the adjustment
        frappe.logger().info(f"Commission adjusted for {invoice_name}: {old_commission} -> {new_commission_amount}")
        
        return {
            "success": True,
            "message": f"Commission adjusted from {frappe.format(old_commission, {'fieldtype': 'Currency'})} to {frappe.format(new_commission_amount, {'fieldtype': 'Currency'})}",
            "old_commission": old_commission,
            "new_commission": new_commission_amount,
            "new_rate": new_commission_rate,
            "outstanding": outstanding
        }
        
    except Exception as e:
        frappe.log_error(f"Error adjusting commission for {invoice_name}: {str(e)}")
        return {
            "success": False,
            "message": str(e)
        }

@frappe.whitelist()
def recalculate_commission_from_rate(invoice_name, commission_rate):
    """Recalculate commission amount from rate"""
    try:
        doc = frappe.get_doc("Sales Invoice", invoice_name)
        base_amount = flt(doc.grand_total)
        
        if doc.dlits_commission_type == "Percentage":
            commission_amount = base_amount * flt(commission_rate) / 100
        else:
            commission_amount = flt(commission_rate)
        
        return {
            "commission_amount": commission_amount,
            "commission_rate": flt(commission_rate)
        }
        
    except Exception as e:
        return {
            "error": str(e)
        }

@frappe.whitelist()
def recalculate_commission_from_amount(invoice_name, commission_amount):
    """Recalculate commission rate from amount"""
    try:
        doc = frappe.get_doc("Sales Invoice", invoice_name)
        base_amount = flt(doc.grand_total)
        
        if base_amount > 0:
            commission_rate = (flt(commission_amount) / base_amount) * 100
        else:
            commission_rate = 0
        
        return {
            "commission_amount": flt(commission_amount),
            "commission_rate": commission_rate
        }
        
    except Exception as e:
        return {
            "error": str(e)
        }